src/components/feedback/Popover.tsx

```diff
@@ -0,0 +1,257 @@
+        return setPopoverCSSProps({
+          bottom: '100%',
+          marginBottom: POPOVER_ANCHOR_EL_GAP,
+        });
+      }
+
+      return setPopoverCSSProps({
+        top: '100%',
+        marginTop: POPOVER_ANCHOR_EL_GAP,
+      });
+    }
+
+    const { top: bodyTop, width: bodyWidth } =
+      document.body.getBoundingClientRect();
+    const absBodyTop = Math.abs(bodyTop);
+
+    // The available space is:
+    // - left-aligned popovers: distance from left side of anchor element to
+    //   right side of viewport
+    // - right-aligned popovers: distance from right side of anchor element to
+    //   left side of viewport
+    const availableSpace =
+      (alignToRight ? anchorElLeft + anchorElWidth : bodyWidth - anchorElLeft) -
+      POPOVER_VIEWPORT_HORIZONTAL_GAP;
+
+    let left = anchorElLeft;
+    if (popoverWidth > availableSpace) {
+      // If the popover is not going to fit the available space, let it "grow"
+      // in the opposite direction
+      left = alignToRight
+        ? POPOVER_VIEWPORT_HORIZONTAL_GAP
+        : left - (popoverWidth - availableSpace);
+    } else if (alignToRight && popoverWidth > anchorElWidth) {
+      // If a right-aligned popover fits the available space, but it's bigger
+      // than the anchor element, move it to the left so that it is aligned with
+      // the right side of the element
+      left -= popoverWidth - anchorElWidth;
+    }
+
+    return setPopoverCSSProps({
+      minWidth: `${anchorElWidth}px`,
+      top: shouldBeAbove
+        ? `calc(${absBodyTop + anchorElDistanceToTop - popoverHeight}px - ${POPOVER_ANCHOR_EL_GAP})`
+        : `calc(${absBodyTop + anchorElDistanceToTop + anchorElHeight}px + ${POPOVER_ANCHOR_EL_GAP})`,
+      left: `${Math.max(POPOVER_VIEWPORT_HORIZONTAL_GAP, left)}px`,
+    });
+  }, [asNativePopover, anchorElementRef, popoverRef, alignToRight]);
+
+  useLayoutEffect(() => {
+    if (!popoverOpen) {
+      return () => {};
+    }
+
+    // First of all, open popover if it's using the native API, otherwise its
+    // size is 0x0 and positioning calculations won't work.
+    const popover = popoverRef.current;
+    if (asNativePopover) {
+      popover?.togglePopover(true);
+    }
+
+    const cleanup = adjustPopoverPositioning();
+
+    if (!asNativePopover) {
+      return cleanup;
+    }
+
+    // Readjust popover position when any element scrolls, just in case that
+    // affected the anchor element position.
+    const listeners = new ListenerCollection();
+    listeners.add(document.body, 'scroll', adjustPopoverPositioning, {
+      capture: true,
+    });
+
+    return () => {
+      if (asNativePopover) {
+        popover?.togglePopover(false);
+      }
+      cleanup();
+      listeners.removeAll();
+    };
+  }, [adjustPopoverPositioning, asNativePopover, popoverOpen, popoverRef]);
+}
+
+export type PopoverProps = {
+  children?: ComponentChildren;
+  classes?: string | string[];
+  variant?: 'panel' | 'custom'; // TODO 'tooltip'
```

Comment:
"Panel" variant adds the style used by `Select`, but it's likely the one we'll use for at-mentions as well.

It has:
* White background
* Box shadow
* Border
* Rounded corners
* Padding